api/src/exam-environment/routes/exam-environment.ts

```diff
@@ -0,0 +1,578 @@
+      ERRORS.FCC_EINVAL_EXAM_ENVIRONMENT_AUTHORIZATION_TOKEN(JSON.stringify(e))
+    );
+  }
+
+  const payload = jwt.decode(encodedToken) as JwtPayload;
+
+  const examEnvironmentAuthorizationToken =
+    payload.examEnvironmentAuthorizationToken;
+
+  const token = await this.prisma.examEnvironmentAuthorizationToken.findFirst({
+    where: {
+      id: examEnvironmentAuthorizationToken
+    }
+  });
+
+  if (!token) {
+    void reply.code(200);
+    return reply.send({
+      data: 'Token does not appear to have been created.'
+    });
+  } else {
+    void reply.code(200);
+    return reply.send({
+      data: {
+        createdDate: token.createdDate
+      }
+    });
+  }
+}
+
+/**
+ * Generates an exam for the user.
+ *
+ * Requires token to be validated and TODO: live longer than the exam attempt.
+ */
+async function postExamGeneratedExamHandler(
+  this: FastifyInstance,
+  req: UpdateReqType<typeof schemas.examEnvironmentPostExamGeneratedExam>,
+  reply: FastifyReply
+) {
+  // Get exam from DB
+  const examId = req.body.examId;
+  // TODO: This throws if `id` is not serializable into an ObjectId
+  //       Should this be caught, and thrown to the client as a client error?
+  const maybeExam = await mapErr(
+    this.prisma.envExam.findUnique({
+      where: {
+        id: examId
+      }
+    })
+  );
+  if (maybeExam.hasError) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeExam.error))
+    );
+  }
+
+  const exam = maybeExam.data;
+
+  if (!exam) {
+    void reply.code(404);
+    return reply.send(
+      ERRORS.FCC_ENOENT_EXAM_ENVIRONMENT_MISSING_EXAM('Invalid exam id given.')
+    );
+  }
+
+  // Check user has completed prerequisites
+  const { user } = req;
+  if (!user) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT('Unreachable. User not authenticated.')
+    );
+  }
+  const isExamPrerequisitesMet = checkPrerequisites(user, true);
+
+  if (!isExamPrerequisitesMet) {
+    void reply.code(403);
+    // TODO: Consider sending unmet prerequisites
+    return reply.send(
+      ERRORS.FCC_EINVAL_EXAM_ENVIRONMENT_PREREQUISITES(
+        'User has not completed prerequisites.'
+      )
+    );
+  }
+
+  // Check user has not completed exam in last 24 hours
+  const maybeExamAttempts = await mapErr(
+    this.prisma.envExamAttempt.findMany({
+      where: {
+        userId: user.id,
+        examId: exam.id
+      }
+    })
+  );
+
+  if (maybeExamAttempts.hasError) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeExamAttempts.error))
+    );
+  }
+
+  const examAttempts = maybeExamAttempts.data;
+
+  const maybeLastAttempt = examAttempts.length
+    ? examAttempts.reduce((latest, current) =>
+        latest.startTimeInMS > current.startTimeInMS ? latest : current
+      )
+    : null;
+
+  if (maybeLastAttempt) {
+    const lastAttempt = maybeLastAttempt;
```

Comment:
```suggestion
  const lastAttempt = examAttempts.length
    ? examAttempts.reduce((latest, current) =>
        latest.startTimeInMS > current.startTimeInMS ? latest : current
      )
    : null;

  if (lastAttempt) {
```
Oh for a real Maybe. Anyway, the type captures the fact its `EnvExamAttempt | null` so we don't need to do anything.